### DISCUSSION TITLE
Invalid Environment Variables: DATABASE_URL 'Invalid url'

### DISCUSSION THREAD

**Discussion by Noah3141** [colinhacks/zod#2806] [Q&A] [answered]
I set up my .env with a local mysql url. It errored because the password had #! in it, which needed to be URL encoded. I got it working. I then set up deployment, by leaving the .env alone, and modifying the schema.prisma file so that instead of using "url = env("DATABASE_URL") it used the direct string to a railway MySQL server.  npx db push to MySQL remote server, got everything working on deployment through vercel by adding env variables to Vercel deployment.
Great
Now I switch back to local, tell prisma to look at local env var instead of railway, so that I can use local mySQL, and it refuses to parse the local env file as it had done just moments before...
![image](https://github.com/colinhacks/zod/assets/66894106/1fc47429-6482-4427-a46a-4f8267300d9a)

![image](https://github.com/colinhacks/zod/assets/66894106/c5b9a58e-7d7e-4c99-bab6-f46ffca15314)
![image](https://github.com/colinhacks/zod/assets/66894106/d6becfee-6ffb-4dfe-8032-c9a21694bced)
![image](https://github.com/colinhacks/zod/assets/66894106/7791498b-8af5-4c6c-a191-37c7a4a8e6e7)

**Comment by Noah3141** [comment]
When all validation is removed from env.mjs so that DATABASE_URL can pass through unvalidated, it starts fine, but then prisma starts saying this:


❌ tRPC failed on user.getAll:
Invalid `prisma.user.findMany()` invocation:


error: Error validating datasource `db`: the URL must start with the protocol `mysql://`.
  -->  schema.prisma:14
   |
13 |     // https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#string
14 |     url      = env("DATABASE_URL")
   |

Validation Error Count: 1


I even copy pasted a DB url from the docs and got this

**Comment by Noah3141** [comment]
Validation Error Count: 1
❌ tRPC failed on user.getAll:
Invalid `prisma.user.findMany()` invocation:


error: Error validating datasource `db`: the URL must start with the protocol `mysql://`.
  -->  schema.prisma:10
   |
 9 |     provider = "mysql"
10 |     url      = env("DATABASE_URL")
   |

Validation Error Count: 1

WHEREAS

![image](https://github.com/colinhacks/zod/assets/66894106/c2194614-90e5-4968-8382-c8f7334c993d)

The line numbers don't even match up????????/

**Comment by Noah3141** [comment]
Upon trying to switch back to SQLite, the same error:

prisma:error 
Invalid `prisma.user.findMany()` invocation:


error: Error validating datasource `db`: the URL must start with the protocol `file:`.
  -->  schema.prisma:10
   |
 9 |     provider = "sqlite"
10 |     url      = env("DATABASE_URL")
   |

Validation Error Count: 1
❌ tRPC failed on user.getAll: 
...

**Comment by JacobWeisenburger** [comment]
Please send a full reproducible code example.

**Comment by Noah3141** [comment]
If I can't even reproduce it in an adjacent environment how could I do that without just sending you the actual codebase

**Comment by JacobWeisenburger** [comment]
if you can't reproduce it in an adjacent environment, how is anyone else supposed to?

try pairing down the issue into the smallest form that it could be and sending that.

Also, this might help:
https://www.youtube.com/watch?v=xlTrDIHhnhs

**Comment by Noah3141** [comment]
I tested the Local mySQL connection url on a second environment. No errors, db calls succeed.
I cloned the repo, and inserted THAT testing URL into it. It works.

When I copy paste the DB url from the broken repo to the cloned repo, it ceases working. Excellent, albeit strange.

```
# DATABASE_URL="mysql://root:uungobuungo8990%21%23@127.0.0.1:3306/new-testament" 
  DATABASE_URL="mysql://root:uungobuungo8990%21%23@127.0.0.1:3306/new-testament"
``` 

In the cloned repo, I can alternate between working and not working by switching between these

**Comment by Noah3141** [comment]
However, when I copy the working URL into the broken repo, it errors again with 'Invalid url"

**Comment by JacobWeisenburger** [comment]
Perhaps you need to upgrade your version of zod?

in zod v3.22.2 this works as expected.

```ts
const schema = z.string().url()
console.log( schema.parse( 'mysql://root:uungobuungo8990%21%23@127.0.0.1:3306/new-testament' ) )
// mysql://root:uungobuungo8990%21%23@127.0.0.1:3306/new-testament
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by Noah3141** [comment]
Good call, but doesn't appear to be it

├── @next-auth/prisma-adapter@1.0.7
├── @prisma/client@5.3.1
├── @t3-oss/env-nextjs@0.6.1
├── @tanstack/react-query@4.35.3
├── @trpc/client@10.38.4
├── @trpc/next@10.38.4
├── @trpc/react-query@10.38.4
├── @trpc/server@10.38.4
├── @types/eslint@8.44.3
├── @types/node@18.17.19
├── @types/react-dom@18.2.7
├── @types/react@18.2.22
├── @typescript-eslint/eslint-plugin@6.7.2
├── @typescript-eslint/parser@6.7.2
├── autoprefixer@10.4.16
├── eslint-config-next@13.5.3
├── eslint@8.50.0
├── next-auth@4.23.1
├── next@13.5.3
├── postcss@8.4.30
├── prettier-plugin-tailwindcss@0.5.4
├── prettier@3.0.3
├── prisma@5.3.1
├── react-dom@18.2.0
├── react-hot-toast@2.4.1
├── react-icons@4.11.0
├── react-loader-spinner@5.4.5
├── react@18.2.0
├── superjson@1.13.1
├── tailwindcss@3.3.3
├── typescript@5.2.2
└── zod@3.22.2

I'm suspecting that truly the error is from @t3-oss/env-nextjs@0.6.1

Being able to alternate the two values as above though is strange, I wonder if some underlying character encoding that I don't understand is hiding a difference.

**Comment by Noah3141** [comment]
Having updated, they now both do not work, each producing


❌ Invalid environment variables: { DATABASE_URL: [ 'Invalid url' ] }
Error: Invalid environment variables
    at T.m (file:///C:/Users/Noah3/Code/test/new-testament-io/node_modules/@t3-oss/env-nextjs/dist/index.mjs:1:500)
    at T (file:///C:/Users/Noah3/Code/test/new-testament-io/node_modules/@t3-oss/env-nextjs/dist/index.mjs:1:690)
    at P (file:///C:/Users/Noah3/Code/test/new-testament-io/node_modules/@t3-oss/env-nextjs/dist/index.mjs:1:1081)
    at file:///C:/Users/Noah3/Code/test/new-testament-io/src/env.mjs:4:20
    at ModuleJob.run (node:internal/modules/esm/module_job:192:25)
    at async DefaultModuleLoader.import (node:internal/modules/esm/loader:228:24)
    at async file:///C:/Users/Noah3/Code/test/new-testament-io/next.config.mjs:5:1

**Comment by JacobWeisenburger** [comment, marked as answer]
Unfortunately, I can't really help if the problem is not Zod. So if you are able to produce code where Zod is not doing the right thing, then I will do my best to help.

**Comment by Noah3141** [comment]
That's alright.
The way this error moves around makes me pretty sure it can't be Zod that's doing it, but rather the additional layer of env processing that t3 is doing
